api/db/changelog/db.changelog-97-user-recent-workspace.xml

```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9
+                    http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+    <changeSet author="als364" id="db.changelog-97-user-recent-workspace">
+        <createTable tableName="user_recent_workspace">
+            <column name="id" type="bigint" autoIncrement="true">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="user_id" type="bigint">
```

Comment:
Oh really? I'd think you could just have an ORM object with those fields and then a helper class for finding the oldest one (and its "slot" like 0, 1, 2, 3). I thought it'd be something like 
```
recentAccess.slot1workspaceId = currrent.workspaceId;
recentAccess.slot1timestamp = currrent.timestamp;
```
and you'd figure out which slot to overwrite by comparing the timestamps, maybe with a multimap from timestamp to slot(s).

But it is a funny pattern. It's just a slight smell that we try to maintain a fixed number of rows in a table matching a key (userId), since that kind of invariant isn't something SQL provides out of the box. OTOH your scheme is better if we wan to change the number of remembered access times.

Note: I don't believe anyone has asserted that the cap at 4 is for security or privacy reasons, because right now the plan is for all accesses to be stored in the workspace audit log.

Reply:
I'd have to make separate functions for updating and deleting each one (or each combination of them if I wanted to batch it). You can't really create queries on the fly with Spring.